File: string/375-e-splitStringAsTwoSubStringWithSameDistinctChars.cpp
#include<string>
#include<iostream>
#include<map>
using STR=std::string;
using MAP=std::map<char,int>;

MAP myCounter(STR s)
{
    MAP ans{};
    for(auto &c:s)
    {
        ans[c]+=1;
    }
    return ans;
}
int split2(STR& s)
{
    if(s.empty())
        return 0;
    MAP c1=myCounter(STR(""));
    MAP c2=myCounter(s);
    int ans=0;
    for(int i=0;i<s.size();++i)
    {
        // STR s1=s.substr(0,i);
        // STR s2=s.substr(i);
        c1[s[i]]+=1;
        c2[s[i]]-=1;
        if(c2[s[i]]==0)
            c2.erase(s[i]);
        if(c1.size()==c2.size())
            ans+=1;

    }
    return ans;
}

int main()
{
    STR s{"aabaa"};
    std::cout<<split2(s);
}

File: string/375-e-splitStringAsTwoSubStringWithSameDistinctChars.py

from collections import Counter

def split2(s):
    if not s:
        return 0
    
    ans=0
    
    for i in range(len(s)):
        c1=Counter(s[:i+1])
        c2=Counter(s[i+1:])
        if len(c1)==len(c2):
            ans+=1
            print(s[:i+1],s[i+1:])
    return ans


def split2Faster(s):
    if not s:
        return 0
    
    ans=0
    
    for i in range(len(s)):
        c1=s[:i+1]
        c2=s[i+1:]
        if set(c1)==set(c2):
            ans+=1
            print(c1,c2)
            
    return ans

def split2Counter(s):
    if not s:
        return 0
    
    c1=Counter()
    c2=Counter(s)
    
    ans=0
    
    for i in range(len(s)-1):
        c1[s[i]]+=1
        c2[s[i]]-=1
        if c2[s[i]]==0:
            del c2[s[i]]
        
        if set(c1.keys())==set(c2.keys()):
            ans+=1
            print(c1,c2)
            
    return ans    
    
print(split2("aabab"))
print(split2Faster("aabaa"))
print(split2Counter("aabab"))